### ISSUE TITLE
Cache sizes and other hardware parameters

### ISSUE THREAD

**Issue by philipturner** [AdaptiveCpp/AdaptiveCpp#929]
I've [reverse-engineered](https://github.com/philipturner/metal-benchmarks) the M1 GPU entirely, and it has some *really* small cache sizes. L1D = 8 KB, L1I = 12 KB, L2 = 256 KB (M1 Pro), 512 KB (M1 Max), 768 KB (base M1, strangely), 1 MB (M1 Ultra). This tradeoff is one of several reasons for unrivaled power efficiency.

The SYCL specification has a `info::device::global_mem_cache_size` query that I wish to implement. Is there any suggestion as to which cache I should use? The L2 is so small and inconsistent that it may be useless to know. Meanwhile, L3 cache ranges in megabytes and is always listed on each chip's Wikipedia page.

**Comment by al42and** [comment]
FYI, in OpenCL, there is no consistency:

- Apple (M1) reports no cache at all
- Intel Arc reports total L2 cache size
- AMD reports L0 cache per compute-unit
- NVIDIA shows L1 per SM (it's configurable; minimal possible value is reported by `clinfo`)

(based on https://opencl.gpuinfo.org/, random poking around, and comparing with https://www.techpowerup.com/)

**Comment by philipturner** [comment]
> AMD reports L0 cache per compute-unit
>
> NVIDIA shows L1 per SM (it's configurable; minimal possible value is reported by clinfo)

Thanks for the information. It seems that most GPUs on the market report the smallest possible cache size. The 8 KB L1 cache would be easy, as it's the same for all Apple GPUs.

**Comment by illuhad** [comment]
Yeah you pretty much have free reign here. The only reason SYCL even has this query is because OpenCL has it (which is not consistent either, as @al42and points out), and SYCL implementations are just expected to return what the backend returns.. I'm not even sure how useful that query even is for real world code. Just return what you think fits best :)
